relay-server/src/actors/project_cache.rs

```diff
@@ -1106,4 +1116,54 @@ mod tests {
             assert!(buffer_rx.try_recv().is_err())
         }
     }
+
+    #[tokio::test]
+    async fn test_eviction() {
+        let num_permits = 5;
+        let buffer_guard: Arc<_> = BufferGuard::new(num_permits).into();
+        let services = mocked_services();
+        let (state_tx, _) = mpsc::unbounded_channel();
+        let (buffer_tx, _) = mpsc::unbounded_channel();
+
+        // Projects should be expired after 2 seconds.
+        let config: Arc<_> = Config::from_json_value(serde_json::json!({
+            "cache": {
+              "project_expiry": 1
+            }
+        }))
+        .unwrap()
+        .into();
+
+        let (mut broker, _) = project_cache_broker_setup(
+            config.clone(),
+            services.clone(),
+            buffer_guard,
+            state_tx,
+            buffer_tx,
+        )
+        .await;
+
+        let key1 = ProjectKey::parse("e12d836b15bb49d7bbf99e64295d995b").unwrap();
+        let project1 = Project::new(key1, config.clone());
+        let key2 = ProjectKey::parse("eeed836b15bb49d7bbf99e64295d9bbb").unwrap();
+        let mut project2 = Project::new(key2, config);
+
+        // Simulate the project state update.
+        project2.get_cached_state(services.project_cache.clone(), true);
+
+        broker.projects.insert(key1, project1);
+        broker.projects.insert(key2, project2);
+
+        // Let's wait for a expiry timeout.
+        tokio::time::sleep(Duration::from_secs(2)).await;
+
+        // One of the project will be removed.
+        broker.evict_stale_project_caches();
+
+        // The `project1` is removed, since the state was never requested, so there is no attempts there
+        // either, and we know it's definitely expired. But we have tried to fetch `project2` and
+        // failed so far, so the backoff is in progress.
+        assert_eq!(broker.projects.len(), 1);
+        assert!(broker.projects.get(&key2).is_some());
```

Comment:
I'm a bit confused by this. Where have we tried requesting project 2 but not project 1?

Reply:
Oh dang, I missed it. Thanks for pointing it out!